### PR TITLE
CORS-2854: azure: Allow users to set visibility to components

### PR DESCRIPTION
**Pull request by rna-afk**
Adding a field in the install config to allow the users to set the visibility of ingress and api servers.

### CODE REVIEW

File: pkg/asset/manifests/ingress.go [hunk 1]
```diff
@@ -135,30 +135,28 @@ func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte,
...
-				Namespace: "openshift-ingress-operator",
-				Name:      "default",
-			},
-			Spec: operatorv1.IngressControllerSpec{
-				EndpointPublishingStrategy: &operatorv1.EndpointPublishingStrategy{
-					Type: operatorv1.LoadBalancerServiceStrategyType,
-					LoadBalancer: &operatorv1.LoadBalancerStrategy{
-						Scope: operatorv1.InternalLoadBalancer,
-					},
+	if config.Publish != types.InternalPublishingStrategy && config.InternalResources != nil && !config.InternalResources.Ingress {
```
Comment (sadasu) [inline pkg/asset/manifests/ingress.go:138]: Could you please add a comment / debug statement explaining the conditions being checked here?

Comment (rna-afk) [reply pkg/asset/manifests/ingress.go:138]: Haven't gotten to testing the ingress private side. I just set it to do the same thing for `publish: Internal` but I'm assuming we might need to do something else here? If that even makes sense?

But the logic here should be if it's either `publish: Internal` or if the ingress is set to true then return nil. Looks like the check should just be `config.InternalResources.Ingress` but I don't think I should be taking this approach anyway.

File: pkg/types/installconfig.go [hunk 2]
```diff
@@ -154,6 +154,9 @@ type InstallConfig struct {
 	// +optional
 	Publish PublishingStrategy `json:"publish,omitempty"`
 
+	// InternalResources controls the visibility of ingress and apiserver. Defaults to public.
+	InternalResources *InternalResources `json:"internal,omitempty"`
```
Comment (patrickdillon) [inline pkg/types/installconfig.go:158]: There's some discussion of this field in [this slack thread](https://redhat-internal.slack.com/archives/C011CSSPBLK/p1698240102565929), where @JoelSpeed suggested using a discriminated union for the `publish` field.  One way we could accomplish this is by adding a third `PublishingStrategy`, `Mixed`, and this field would only be configurable if `Mixed` publishing strategy is set.

Comment (patrickdillon) [inline pkg/types/installconfig.go:158]: We should brainstorm a bit on the naming here `InternalResources` and `internal` are a bit vague. I think the effect here is we are distinguishing the publishing strategy for individual cluster components, so we could do something like `OperatorPublishingStrategy`, which would be distinguished from the previous publishing strategy which was for the cluster as a whole.

Comment (rna-afk) [reply pkg/types/installconfig.go:158]: Yeah that's what I have to do, if I can only get this thing to work

File: data/data/azure/vnet/public-lb.tf [hunk 3]
```diff
@@ -5,9 +5,9 @@ locals {
 
 locals {
   // DEBUG: Azure apparently requires dual stack LB for v6
-  need_public_ipv4 = ! var.azure_private || var.azure_outbound_routing_type != "UserDefinedRouting"
+  need_public_ipv4 = ! var.azure_private || ! var.azure_lb_private || var.azure_outbound_routing_type != "UserDefinedRouting"
```
Comment (patrickdillon) [inline data/data/azure/vnet/public-lb.tf:8]: Do we need a second `azure_lb_private` variable in addition to `azure_private`? It looks like it is the same as `azure_private` which would make sense to me: considering that Terraform only creates the API load balancers, I'm not sure we would need any changes to the terraform configs at all. I think we would only need the distinction if Terraform creates any resources related to ingress (or the registry if we're including that component) such as security groups.

Comment (rna-afk) [reply data/data/azure/vnet/public-lb.tf:8]: I was having this issue where terraform wasn't allowing me to associate the bootstrap NIC that has a public IP with a LB that has an outbound rule. I was trying to create a new NIC card for bootstrap only when the lb is private so I needed this variable. I think I need this for other reasons like converting the cname record to A-record on the DNS if lb is private.

Comment (patrickdillon) [reply data/data/azure/vnet/public-lb.tf:8]: > I was having this issue where terraform wasn't allowing me to associate the bootstrap NIC that has a public IP with a LB that has an outbound rule.

Doesn't this already work when the publishing strategy is set to Internal or private?

Unless there are connections/overlap in the terraform configs with ingress/registry I don't see why we would need changes to Terraform, which should really simplify things.

Comment (rna-afk) [reply data/data/azure/vnet/public-lb.tf]: It works in internal if the bootstrap is also private. If the bootstrap is public and the lb is private we get into some trouble.

Comment (patrickdillon) [reply data/data/azure/vnet/public-lb.tf]: > If the bootstrap is public and the lb is private we get into some trouble.


I don't think that is a valid config. If the cluster is private/internal bootstrap should not have a public IP.

https://github.com/openshift/installer/blob/9ff8208c93feb0569b77bf583c85087b8d638c44/data/data/azure/bootstrap/main.tf#L79

Comment (rna-afk) [reply data/data/azure/vnet/public-lb.tf]: But the requirement is to keep just the apiserver private and the others public.

Comment (patrickdillon) [reply data/data/azure/vnet/public-lb.tf]: The bootstrap node's not a separate component like ingress or the registry--visibility should be the same as the apiserver in all cases. If there is indeed a specific requirement to have bootstrap visibility configurable as separate from the API server LMK.

**Comment by rna-afk** [comment]
Added a small change to baseDomain validation to include MixedPublishingStrategy.

File: pkg/types/validation/installconfig.go [hunk 4]
```diff
@@ -155,6 +155,35 @@ func ValidateInstallConfig(c *types.InstallConfig, usingAgentMethod bool) field.
 		}
 	}
 
+	if c.Publish == types.MixedPublishingStrategy {
+		switch platformName := c.Platform.Name(); platformName {
+		case azure.Name:
+		default:
```
Comment (r4f4) [inline pkg/types/validation/installconfig.go:161]: Nit: Maybe `if` looks better here?
```suggestion
		if platformName := c.Platform.Name(); platformName != azure.Name {
```

File: pkg/types/validation/installconfig.go [hunk 5]
```diff
@@ -155,6 +155,35 @@ func ValidateInstallConfig(c *types.InstallConfig, usingAgentMethod bool) field.
...
+			allErrs = append(allErrs, field.Invalid(field.NewPath("publish"), c.Publish, fmt.Sprintf("mixed publish strategy is not supported on %q platform", platformName)))
+		}
+		if c.OperatorPublishingStrategy == nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("publish"), c.Publish, "please specify the operator publishing strategy for mixed publish strategy"))
+		}
```
Comment (r4f4) [inline pkg/types/validation/installconfig.go:166]: So you're forced to specify a value even if the default is `External`?

File: pkg/types/validation/installconfig.go [hunk 6]
```diff
@@ -155,6 +155,35 @@ func ValidateInstallConfig(c *types.InstallConfig, usingAgentMethod bool) field.
...
+	} else if c.OperatorPublishingStrategy != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("operatorPublishingStrategy"), c.Publish, "operator publishing strategy is only allowed with mixed publishing strategy installs"))
+	}
+
+	if c.OperatorPublishingStrategy != nil {
+		acceptedValues := sets.NewString("Internal", "External")
```
Comment (r4f4) [inline pkg/types/validation/installconfig.go:172]: Let's use the non-deprecated set syntax for new code
```suggestion
		acceptedValues := sets.New[string]("Internal", "External")
```

File: pkg/types/validation/installconfig.go [hunk 7]
```diff
@@ -155,6 +155,35 @@ func ValidateInstallConfig(c *types.InstallConfig, usingAgentMethod bool) field.
...
+		if c.OperatorPublishingStrategy.APIServer == "" {
+			c.OperatorPublishingStrategy.APIServer = "External"
+		}
+		if c.OperatorPublishingStrategy.Ingress == "" {
+			c.OperatorPublishingStrategy.Ingress = "External"
+		}
+		if !acceptedValues.Has(c.OperatorPublishingStrategy.APIServer) {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("apiserver"), c.OperatorPublishingStrategy.APIServer, acceptedValues.List()))
```
Comment (r4f4) [inline pkg/types/validation/installconfig.go:180]:
```suggestion
			allErrs = append(allErrs, field.NotSupported(field.NewPath("apiserver"), c.OperatorPublishingStrategy.APIServer, sets.List(acceptedValues)))
```

File: pkg/types/validation/installconfig.go [hunk 8]
```diff
@@ -155,6 +155,35 @@ func ValidateInstallConfig(c *types.InstallConfig, usingAgentMethod bool) field.
...
+		}
+		if !acceptedValues.Has(c.OperatorPublishingStrategy.Ingress) {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("ingress"), c.OperatorPublishingStrategy.Ingress, acceptedValues.List()))
```
Comment (r4f4) [inline pkg/types/validation/installconfig.go:183]:
```suggestion
			allErrs = append(allErrs, field.NotSupported(field.NewPath("ingress"), c.OperatorPublishingStrategy.Ingress, sets.List(acceptedValues))
```

File: pkg/types/validation/installconfig.go (hunk 5 above)
Comment (r4f4) [reply pkg/types/validation/installconfig.go:166]: Ok, fair enough

**Review by r4f4 — COMMENTED** [review]
/lgtm

File: pkg/types/azure/validation/platform.go [hunk 9]
```diff
@@ -52,7 +52,7 @@ func ValidatePlatform(p *azure.Platform, publish types.PublishingStrategy, fldPa
 	if p.Region == "" {
 		allErrs = append(allErrs, field.Required(fldPath.Child("region"), "region should be set to one of the supported Azure regions"))
 	}
-	if !p.IsARO() && publish == types.ExternalPublishingStrategy {
+	if !p.IsARO() && publish != types.InternalPublishingStrategy {
```
Comment (jinyunma) [inline pkg/types/azure/validation/platform.go:55]: Nit: actually, publish: Mixed + apiserver: Internal + ingress: Internal is the similar scenario as private cluster (publish: Internal), it also doesn't require BaseDomainResourceGroupName. Do we need to consider it?

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Comment by patrickdillon** [comment]
/retest-required

pods are timing out

**Comment by rna-afk** [comment]
Rebased after #7520 merged

**Pull request merged**